File: src/main/java/com/maxdreher/table/TableEntry.java
package com.maxdreher.table;

import android.content.Context;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.maxdreher.consumers.ConsumeAndSupply;
import com.maxdreher.consumers.ConsumeTwo;
import com.maxdreher.consumers.ConsumeTwoAndSupply;

import java.util.Comparator;

//TODO #6
public class TableEntry<Input> {

    private final String name;
    private Comparator<Input> comparator;
    private final ConsumeTwoAndSupply<Context, Input, ? super View> generate;
    private final int weight;


    public TableEntry(String name, Comparator<Input> comparator, ConsumeTwoAndSupply<Context, Input, ? super View> generator) {
        this(name, comparator, 1, generator);
    }

    public TableEntry(String name, Comparator<Input> comparator, int weight, ConsumeTwoAndSupply<Context, Input, ? super View> generator) {
        this.name = name;
        this.comparator = comparator;
        this.weight = weight;
        this.generate = generator;
    }

    public String getName() {
        return name;
    }

    public boolean isSortable() {
        return comparator != null;
    }

    public Comparator<Input> getComparator() {
        return comparator;
    }

    public View generate(Context context, Input input) {
        return (View) generate.consume(context, input);
    }


    public static <Input> ConsumeTwoAndSupply<Context, Input, View> textViewGenerator(ConsumeAndSupply<Input, String> toText, int marginsPixels) {
        return textViewGenerator(toText, null, marginsPixels);
    }

    public static <Input> ConsumeTwoAndSupply<Context, Input, View> textViewGenerator(ConsumeAndSupply<Input, String> toText, ConsumeTwo<TextView, Input> viewConsumer, int marginsPixels) {
        return (context, input) -> {
            LinearLayout linearLayout = new LinearLayout(context);
            linearLayout.setOrientation(LinearLayout.HORIZONTAL);
            linearLayout.setGravity(Gravity.CENTER);
            linearLayout.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

            TextView textView = new TextView(context);
            textView.setText(toText.consume(input));
            textView.setGravity(Gravity.CENTER);

            if (marginsPixels != -1) {
                textView.setPadding(marginsPixels, marginsPixels, marginsPixels, marginsPixels);
            }
            if (viewConsumer != null) {
                viewConsumer.consume(textView, input);
            }

            linearLayout.addView(textView);
            return linearLayout;
        };
    }

    public int getWeight() {
        return weight;
    }
}

File: src/main/java/com/maxdreher/table/TableEntry.kt
package com.maxdreher.table

import android.content.Context
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import com.maxdreher.consumers.ConsumeAndSupply
import com.maxdreher.consumers.ConsumeTwo
import com.maxdreher.consumers.ConsumeTwoAndSupply
import de.codecrafters.tableview.SortableTableView
import java.util.*


/**
 * A table value for a [SortableTableView]
 *
 * Holds:
 * name of column
 * a converter from input -> view (can be assisted with [textViewGenerator])
 * comparator
 * layout weight
 */
class TableEntry<Input> @JvmOverloads constructor(
    val name: String,
    private val viewGenerator: (Context, Input) -> View,
    val weight: Int = 1,
    val comparator: Comparator<Input>? = null,
) {

    val isSortable: Boolean
        get() = comparator != null

    fun generate(context: Context, input: Input): View {
        return viewGenerator.invoke(context, input)
    }

    companion object {

        fun <Input> from(
            margin: Int,
            map: Map<String, (Input) -> String>
        ): List<TableEntry<Input>> {
            return map.map { toText(it.key, it.value, margin) }
        }

        fun <Input> toText(
            name: String,
            toString: (Input) -> String,
            margin: Int
        ): TableEntry<Input> {
            return TableEntry(name, textViewGenerator(toString, margin))
        }

        fun <Input> textViewGenerator(
            convertToText: (Input) -> String,
            margin: Int,
            viewConsumer: ((TextView, Input) -> Unit)? = null
        ): (Context, Input) -> View {
            return { context: Context, input: Input ->

                val linearLayout = LinearLayout(context)

                linearLayout.apply {
                    orientation = LinearLayout.HORIZONTAL
                    gravity = Gravity.CENTER
                    layoutParams = LinearLayout.LayoutParams(
                        ViewGroup.LayoutParams.MATCH_PARENT,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                    )
                }

                val textView = TextView(context).apply {
                    text = convertToText.invoke(input)
                    gravity = Gravity.CENTER
                }

                if (margin != -1) {
                    textView.setPadding(margin, margin, margin, margin)
                }

                viewConsumer?.invoke(textView, input)

                linearLayout.addView(textView)
                linearLayout
            }
        }
    }
}